[R-package] fix protection stack imbalance and unprotected objects (fixes #4390)

fabsig opened the pull request:
Fixes  #4390.

jameslamb reviewed (changes requested):
Thank you very much for your efforts and clear description of the problem in #4390!

I'll be able to give this a more thorough review later today or maybe tomorrow. For now, could you please address the minor linting issues described in https://github.com/microsoft/LightGBM/pull/4391/checks?check_run_id=2858695056?

You can see the following to understand how to replicate those locally:

https://github.com/microsoft/LightGBM/blob/42ddff1dec813a868478e698f4fe3883dfde61b6/.ci/test.sh#L66

https://github.com/microsoft/LightGBM/blob/42ddff1dec813a868478e698f4fe3883dfde61b6/.ci/test.sh#L75

jameslamb commented:
/gha run r-valgrind

jameslamb commented:
/gha run r-solaris

fabsig commented:
Linting issues should be fixed.

jameslamb commented:
Thanks! Sorry for the delays in CI runs...GitHub has a new mechanism (introduced to prevent abuse) which requires maintainers to manually approve CI runs for new contributors.

![image](https://user-images.githubusercontent.com/7608904/122575344-43c87280-d016-11eb-95a3-9578ef7d90e7.png)

fabsig commented:
No hurry. A few comments on my changes that might be helpful for easier understanding:

- I dropped all return statements in the catch statements in the `R_API_END` and `CHECK_CALL` macros to prevent potential protection stack imbalances. I don't see any use of these return statements.
- I wrapped the `R_API_BEGIN` and `R_API_END` only around the calls to the C functions. It seems to me that the other operations don't need any error handling. In this way, the exception handling is encapsulated away from the protection
- Added protection where needed

jameslamb commented:
@fabsig we use a continuous integration job triggered by pull request comments which replicates CRAN's use of `valgrind` to check incoming packages.

I've tried to run it twice on this PR but the job is timing out.

https://github.com/microsoft/LightGBM/actions/runs/950064363

> The job running on runner GitHub Actions 23 has exceeded the maximum execution time of 120 minutes.

It's possible that we were already very close to the 120-minute limit and that the changes here added enough time to hit that limit. Could you please update https://github.com/microsoft/LightGBM/blob/c7134fa7ccbe1a1b1819ac2a66407504d3fab27a/.github/workflows/r_valgrind.yml#L10 to 180 in this PR? Just so we can see how long it takes.

Really sorry for the inconvenience.

fabsig commented:
> Could you please update
> 
> https://github.com/microsoft/LightGBM/blob/c7134fa7ccbe1a1b1819ac2a66407504d3fab27a/.github/workflows/r_valgrind.yml#L10


Of course. Let me know if further changes are needed.

jameslamb commented:
/gha run r-valgrind

jameslamb commented on R-package/src/lightgbm_R.cpp:
```diff
@@ -77,97 +80,111 @@ SEXP LGBM_DatasetCreateFromCSC_R(SEXP indptr,
   int32_t len = static_cast<int32_t>(Rf_asInteger(len_used_row_indices));
   std::vector<int32_t> idxvec(len);
-  // convert from one-based to zero-based index
+  // convert from one-based to  zero-based index
```
```suggestion
  // convert from one-based to zero-based index
```

This extra space looks like a mistake, could you please revert it?

jameslamb commented on R-package/src/lightgbm_R.cpp:
```diff
@@ -284,29 +309,37 @@ SEXP LGBM_DatasetGetField_R(SEXP handle,
-  CHECK_CALL(LGBM_DatasetUpdateParamChecking(CHAR(Rf_asChar(old_params)), CHAR(Rf_asChar(new_params))));
+  CHECK_CALL(LGBM_DatasetUpdateParamChecking(old_params_ptr, new_params_ptr));
   R_API_END();
+  UNPROTECT(2);
```
In some places in this PR, you chose to increment a counter `n_protected` and then call `UNPROTECT(n_protected)`. In other places, you used code like this that just hard-codes a number.

Is there a reason for that? If not, I'd like to use a consistent pattern throughout this file. I have a mild preference for the pattern where you use an integer literal like `UNPROTECT(2)`, just for code conciseness. I think we could rely on a CI job soon to catch mistakes of the form "imbalance of UNPROTECT() and PROTECT() calls" (#4400).

fabsig replied:
Good point. No particular reason... I thought about that myself and agree with your assessment. I will change this.

jameslamb reviewed (changes requested):
Thanks very much for this!

I ran the following to run `rchk` on this branch, and can confirm these changes eliminate the stack imbalance and protection errors found on `master` (https://github.com/microsoft/LightGBM/issues/4390#issuecomment-866527566) :tada: 

```shell
git clone --recursive git@github.com:fabsig/LightGBM.git fabsig-LightGBM
cd fabsig-LightGBM
git fetch origin R_package_protection
git checkout R_package_protection

sh build-cran-package.sh
mkdir packages
cp lightgbm_3.2.1.99.tar.gz packages
docker run -v `pwd`/packages:/rchk/packages kalibera/rchk:latest /rchk/packages/lightgbm_3.2.1.99.tar.gz
```

end of `rchk` logs on this branch

```text
Installed libraries of package  lightgbm 
[1] "/rchk/packages/libsonly/lightgbm/libs/lightgbm.so"

Library name (usually package name): lightgbm
Initialization function: R_init_lightgbm
Functions: 41
Checked call to R_registerRoutines: 1
ERROR: too many states (abstraction error?) in function strptime_internal
Analyzed 39640 functions, traversed 107321 states.

Rchk version: 3d653b7c8f92dac912532856b55f44d2986c6553
R version: 79889/R Under development (unstable) (2021-01-27 r79889)
LLVM version: 10.0.0
```

So overall, this looks great! Please just see a few minor comments I left about formatting.

<hr>

I think that our friends over at `{xgboost}` (@trivialfis @hcho3) would also really appreciate a similar contribution, if you have time. That project has similar uses of inline `CHAR(asChar())` without protection. For example:

https://github.com/dmlc/xgboost/blob/da1ad798cabb6868be368bd262814cebcbf3d094/R-package/src/xgboost_R.cc#L416-L421

Would you consider opening an issue similar to #4390 in https://github.com/dmlc/xgboost/issues? Not strictly required here, but we have borrowed a lot from that project and try to give back to it when possible.

fabsig commented:
> I think that our friends over at `{xgboost}` (@trivialfis @hcho3) would also really appreciate a similar contribution, if you have time. That project has similar uses of inline `CHAR(asChar())` without protection. For example:
> 
> https://github.com/dmlc/xgboost/blob/da1ad798cabb6868be368bd262814cebcbf3d094/R-package/src/xgboost_R.cc#L416-L421
> 
> Would you consider opening an issue similar to #4390 in https://github.com/dmlc/xgboost/issues? Not strictly required here, but we have borrowed a lot from that project and try to give back to it when possible.

Thanks for the suggestion. I will do that.

trivialfis commented:
Thanks for letting us know!

jameslamb commented on .github/workflows/r_valgrind.yml:
```diff
@@ -7,7 +7,7 @@ on:
   test-r-valgrind:
     name: r-package (ubuntu-latest, R-devel, valgrind)
-    timeout-minutes: 120
+    timeout-minutes: 180
```
```suggestion
    timeout-minutes: 120
```

Apologies, I misunderstood how this check works. This config is sourced from `master` when the relevant CI job runs, so changing it here won't work.

https://github.com/microsoft/LightGBM/pull/4397#issuecomment-866790489

Once #4404 is merged, I'll test this again.

jameslamb reviewed (changes requested):
These changes look great! Thanks for a narrowly-scoped and very clearly-explained pull request that was easy to review.

I am leaving a "request changes" review only to block merging for now. I want to wait to merge this until we can run `valgrind` checks on this branch, which won't be possible until #4404 is merged.

StrikerRUS commented:
/gha run r-valgrind

jameslamb reviewed (approved):
`valgrind` seems happy with these changes (https://github.com/microsoft/LightGBM/pull/4391#issuecomment-868993166).

I also re-ran `rchk` on this branch and it did not detect any remaining issues.

```text
Installed libraries of package  lightgbm
[1] "/rchk/packages/libsonly/lightgbm/libs/lightgbm.so"
/rchk/scripts/check_package.sh: line 100:   561 Killed                  $RCHK/src/$T $RBC $F > $FOUT 2>&1

Library name (usually package name): lightgbm
Initialization function: R_init_lightgbm
Functions: 41
Checked call to R_registerRoutines: 1

Rchk version: 3d653b7c8f92dac912532856b55f44d2986c6553
R version: 79889/R Under development (unstable) (2021-01-27 r79889)
LLVM version: 10.0.0
```

Thanks very much for this contribution 🙌
